### PR TITLE
Time Gate: Selectable gating methods

### PR DESCRIPTION
**Pull request by Vinc0110**
This PR refactors `time.time_gate()` to perform the gating either by convolution, by FFT (band-pass mode), or by RFFT (low-pass mode).

A new parameter `method` is added, which defaults to FFT.

Also, another parameter `fft_window` is added, which enables frequency-domain windowing in case of FFT and RFFT mode. The default is a Hann window, but windowing can be disabled by setting `fft_window=None`.

### CODE REVIEW

**Comment by Vinc0110** [comment]
@RA226, I cannot request your review because you have not yet contributed to this respository, but please have a look at the code.

**Review by jhillairet — APPROVED** [review]
That's great. 

Do you think you could make a new notebook dedicated to this question? using the different tests and comparisons made in the issue thread?

**Comment by Vinc0110** [comment]
Sure. But I would like to wait until the issue with the strange roll-off is solved. Once everything is clear, I will write the notebook.

**Comment by Vinc0110** [comment]
Not sure how to proceed here. As Roger mentioned in the issue thread, he currently does not have time to review the problem. I myself am not a huge expert on the intricacies of FFT and alike, so I can only wait for new ideas.

We could leave this open and not include it in the next release. Or we just merge this for now and implement changes later if required.

**Review by mhuser — APPROVED** [review]
Excellent PR! I like the fact it offer various implementation alternatives for the advanced user.

**Comment by Vinc0110** [comment]
@mhuser Thanks for the suggestions. I'll work on that soon and run a few more tests.

File: skrf/time.py [hunk 1]
```diff
@@ -268,32 +295,82 @@ def time_gate(ntwk: 'Network', start: float = None, stop: float = None, center:
...
     gate[start_idx:stop_idx] = window
 
-    # time-domain gating
-    s_td = fftshift(ifft(nw_gated.s[:, 0, 0]))
-    s_td_g = s_td * gate
-    nw_gated.s[:, 0, 0] = fft(ifftshift(s_td_g))
+    if method == 'convolution':
+        # frequency-domain gating
+        kernel = fftshift(fft(ifftshift(gate), norm='forward'))
+        ntwk_gated.s[:, 0, 0] = convolve1d(ntwk_gated.s[:, 0, 0], kernel, mode='wrap')
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/time.py]: I tried `mode=constant`. It is slightly better, but still not perfect. `cval=0` is the default, so I added a new parameter for the user to select the convolution mode. It looks like the right choice depends on the actual data.

File: skrf/time.py [hunk 2]
```diff
@@ -268,32 +295,82 @@ def time_gate(ntwk: 'Network', start: float = None, stop: float = None, center:
...
+    elif method == 'fft':
+        # time-domain band-pass mode
+        s_td = fftshift(ifft(ntwk_gated.s[:, 0, 0]))
+        s_td_g = s_td * gate
+        ntwk_gated.s[:, 0, 0] = fft(ifftshift(s_td_g))
+    elif method == 'rfft':
+        # time-domain low-pass mode
+        s_td = fftshift(irfft(ntwk_gated.s[:, 0, 0], n=len(t)))
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/time.py:376]: Yes. I added a warning.

**Comment by Vinc0110** [comment]
This is ready from my side. If Roger (@RA226) has more ideas to deal with boundary effects, we should probably do it in another PR.
Feel free to merge if this is ok.

**Review by mhuser — APPROVED** [review]
Great PR!

**Review by jhillairet — APPROVED** [review]

**Comment by RA226** [comment]
I am swamped with other things now. I will look at this again as soon as I can. If I have anything useful to consider it could be included in another PR.

**Pull request merged**